File: javascript/test/encoder.test.js
'use strict';

const test = require('node:test');
const assert = require('node:assert/strict');

const { numberToId } = require('../src/encoder');
const { BLACKLIST, MAX_NUM } = require('../src/constants');

test('numberToId(0) returns a valid 4-syllable id', () => {
  const id = numberToId(0);
  assert.match(id, /^[A-Z][a-z]{3} [A-Z][a-z]{3}$/);
});

test('numberToId(1) differs from numberToId(0)', () => {
  assert.notEqual(numberToId(1), numberToId(0));
});

test('numberToId(10000) matches reference output', () => {
  assert.equal(numberToId(10000), 'Talo Buno');
});

test('numberToId(10001) matches reference output', () => {
  assert.equal(numberToId(10001), 'Patu Luta');
});

test('numberToId(MAX_NUM) succeeds and matches reference output', () => {
  assert.equal(numberToId(MAX_NUM), 'Dobu Zusa');
});

test('numberToId(MAX_NUM + 1) throws', () => {
  assert.throws(() => numberToId(MAX_NUM + 1), RangeError);
});

test('numberToId(-1) throws', () => {
  assert.throws(() => numberToId(-1), RangeError);
});

test('numberToId(1.5) throws', () => {
  assert.throws(() => numberToId(1.5), RangeError);
});

test('capitalization: first letter uppercase only', () => {
  const id = numberToId(42);
  const [w1, w2] = id.split(' ');
  assert.equal(w1, w1[0].toUpperCase() + w1.slice(1).toLowerCase());
  assert.equal(w2, w2[0].toUpperCase() + w2.slice(1).toLowerCase());
});

test('determinism: same input always produces the same output', () => {
  const n = 987654;
  assert.equal(numberToId(n), numberToId(n));
});

test('uniqueness and blacklist-avoidance across the entire valid range', () => {
  const seen = new Set();
  for (let n = 0; n <= MAX_NUM; n += 1) {
    const id = numberToId(n);
    assert.ok(!seen.has(id), `${id} (from ${n}) was already issued for a different number`);
    seen.add(id);

    const [w1, w2] = id.toLowerCase().split(' ');
    assert.ok(!BLACKLIST.has(w1), `${w1} from ${n} is blacklisted`);
    assert.ok(!BLACKLIST.has(w2), `${w2} from ${n} is blacklisted`);
  }
  assert.equal(seen.size, MAX_NUM + 1);
});

File: java/src/test/java/com/javax0/babbler/EncoderTest.java
package com.javax0.babbler;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.HashSet;
import java.util.Set;
import java.util.regex.Pattern;
import org.junit.jupiter.api.Test;

class EncoderTest {

    private static final Pattern ID_PATTERN = Pattern.compile("^[A-Z][a-z]{3} [A-Z][a-z]{3}$");
    private static final IdEncoder encoder = new Encoder();

    @Test
    void numberToIdZeroReturnsValidId() {
        assertTrue(ID_PATTERN.matcher(encoder.numberToId(0)).matches());
    }

    @Test
    void numberToIdOneDiffersFromZero() {
        assertNotEquals(encoder.numberToId(0), encoder.numberToId(1));
    }

    @Test
    void numberToId10000MatchesReference() {
        assertEquals("Talo Buno", encoder.numberToId(10000));
    }

    @Test
    void numberToId10001MatchesReference() {
        assertEquals("Patu Luta", encoder.numberToId(10001));
    }

    @Test
    void numberToIdMaxNumMatchesReference() {
        assertEquals("Dobu Zusa", encoder.numberToId((int) Constants.MAX_NUM));
    }

    @Test
    void numberToIdOutOfRangeThrows() {
        assertThrows(IllegalArgumentException.class, () -> encoder.numberToId((int) Constants.MAX_NUM + 1));
    }

    @Test
    void numberToIdNegativeThrows() {
        assertThrows(IllegalArgumentException.class, () -> encoder.numberToId(-1));
    }

    @Test
    void capitalizationFirstLetterOnly() {
        String[] parts = encoder.numberToId(42).split(" ");
        for (String word : parts) {
            assertEquals(Character.toUpperCase(word.charAt(0)) + word.substring(1).toLowerCase(), word);
        }
    }

    @Test
    void determinism() {
        assertEquals(encoder.numberToId(987654), encoder.numberToId(987654));
    }

    @Test
    void uniquenessAndBlacklistAvoidanceAcrossEntireValidRange() {
        Set<String> seen = new HashSet<>();
        for (int n = 0; n <= Constants.MAX_NUM; n++) {
            String id = encoder.numberToId(n);
            assertTrue(seen.add(id), id + " (from " + n + ") was already issued for a different number");

            String[] words = id.toLowerCase().split(" ");
            assertFalse(Constants.BLACKLIST.contains(words[0]), words[0] + " from " + n + " is blacklisted");
            assertFalse(Constants.BLACKLIST.contains(words[1]), words[1] + " from " + n + " is blacklisted");
        }
        assertEquals(Constants.MAX_NUM + 1, seen.size());
    }
}
